authority/provisioner/oidc.go

```diff
@@ -0,0 +1,205 @@
+	Admins                []string `json:"admins,omitempty"`
+	configuration         openIDConfiguration
+	keyStore              *keyStore
+}
+
+// IsAdmin returns true if the given email is in the Admins whitelist, false
+// otherwise.
+func (o *OIDC) IsAdmin(email string) bool {
+	for _, e := range o.Admins {
+		if e == email {
+			return true
+		}
+	}
+	return false
+}
+
+// GetID returns the provisioner unique identifier, the OIDC provisioner the
+// uses the clientID for this.
+func (o *OIDC) GetID() string {
+	return o.ClientID
+}
+
+// GetName returns the name of the provisioner.
+func (o *OIDC) GetName() string {
+	return o.Name
+}
+
+// GetType returns the type of provisioner.
+func (o *OIDC) GetType() Type {
+	return TypeOIDC
+}
+
+// GetEncryptedKey is not available in an OIDC provisioner.
+func (o *OIDC) GetEncryptedKey() (kid string, key string, ok bool) {
+	return "", "", false
+}
+
+// Init validates and initializes the OIDC provider.
+func (o *OIDC) Init(config Config) (err error) {
+	switch {
+	case o.Name == "":
+		return errors.New("name cannot be empty")
+	case o.ClientID == "":
+		return errors.New("clientID cannot be empty")
+	case o.ConfigurationEndpoint == "":
+		return errors.New("configurationEndpoint cannot be empty")
+	}
+
+	// Update claims with global ones
+	if o.Claims, err = o.Claims.Init(&config.Claims); err != nil {
+		return err
+	}
+	// Decode and validate openid-configuration endpoint
+	if err := getAndDecode(o.ConfigurationEndpoint, &o.configuration); err != nil {
+		return err
+	}
+	if err := o.configuration.Validate(); err != nil {
+		return errors.Wrapf(err, "error parsing %s", o.ConfigurationEndpoint)
+	}
+	// Get JWK key set
+	o.keyStore, err = newKeyStore(o.configuration.JWKSetURI)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidatePayload validates the given token payload.
+func (o *OIDC) ValidatePayload(p openIDPayload) error {
+	// According to "rfc7519 JSON Web Token" acceptable skew should be no more
+	// than a few minutes.
+	if err := p.ValidateWithLeeway(jose.Expected{
+		Issuer:   o.configuration.Issuer,
+		Audience: jose.Audience{o.ClientID},
+		Time:     time.Now().UTC(),
+	}, time.Minute); err != nil {
+		return errors.Wrap(err, "failed to validate payload")
+	}
+	if p.AuthorizedParty != "" && p.AuthorizedParty != o.ClientID {
+		return errors.New("failed to validate payload: invalid azp")
+	}
+	return nil
+}
+
+// Authorize validates the given token.
+func (o *OIDC) Authorize(token string) ([]SignOption, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing token")
+	}
+
+	// Parse claims to get the kid
+	var claims openIDPayload
+	if err := jwt.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return nil, errors.Wrap(err, "error parsing claims")
+	}
+
+	found := false
+	kid := jwt.Headers[0].KeyID
+	keys := o.keyStore.Get(kid)
+	for _, key := range keys {
```

Comment:
This could use a comment. As far as I can tell you're using the first element in the keyset that successfully marshals into the claims object you want.